hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/RenameAtomicity.java

```diff
@@ -0,0 +1,380 @@
+      final String srcEtag,
+      final AbfsClient abfsClient) throws IOException {
+    this.src = src;
+    this.dst = dst;
+    this.abfsClient = (AbfsBlobClient) abfsClient;
+    this.renameJsonPath = renameJsonPath;
+    this.tracingContext = tracingContext;
+    this.srcEtag = srcEtag;
+  }
+
+  /**
+   * Resumes the rename operation from the JSON file.
+   *
+   * @param renameJsonPath Path of the JSON file
+   * @param tracingContext Tracing context
+   * @param srcEtag ETag of the source directory
+   * @param abfsClient AbfsClient instance
+   * @throws IOException Server error while reading / deleting json file.
+   */
+  public RenameAtomicity(final Path renameJsonPath,
+      final int renamePendingJsonFileLen,
+      TracingContext tracingContext,
+      final String srcEtag,
+      final AbfsClient abfsClient) {
+    this.abfsClient = (AbfsBlobClient) abfsClient;
+    this.renameJsonPath = renameJsonPath;
+    this.tracingContext = tracingContext;
+    this.srcEtag = srcEtag;
+    this.renamePendingJsonLen = renamePendingJsonFileLen;
+  }
+
+  public void redo() throws IOException {
+    // parse the JSON
+    byte[] buffer = readRenamePendingJson(renameJsonPath, renamePendingJsonLen);
+    String contents = new String(buffer, Charset.defaultCharset());
+    JsonNode oldFolderName, newFolderName, eTag;
+    try {
+      try {
+        JsonNode json = READER.readValue(contents);
+
+        // initialize this object's fields
+        oldFolderName = json.get("OldFolderName");
+        newFolderName = json.get("NewFolderName");
+        eTag = json.get("ETag");
+      } catch (JsonMappingException | JsonParseException e) {
+        this.src = null;
+        this.dst = null;
+        this.srcEtag = null;
+        return;
+      }
+      if (oldFolderName != null && StringUtils.isNotEmpty(
+          oldFolderName.textValue())
+          && newFolderName != null && StringUtils.isNotEmpty(
+          newFolderName.textValue()) && eTag != null
+          && StringUtils.isNotEmpty(
+          eTag.textValue())) {
+        this.src = new Path(oldFolderName.asText());
+        this.dst = new Path(newFolderName.asText());
+        this.srcEtag = eTag.asText();
+
+        BlobRenameHandler blobRenameHandler = new BlobRenameHandler(
+            this.src.toUri().getPath(), dst.toUri().getPath(),
+            abfsClient, srcEtag, true, true, tracingContext);
+
+        blobRenameHandler.execute();
+      } else {
+        this.src = null;
+        this.dst = null;
+        this.srcEtag = null;
+      }
+    } finally {
+      deleteRenamePendingJson();
+    }
+  }
+
+  @VisibleForTesting
+  byte[] readRenamePendingJson(Path path, int len)
+      throws AzureBlobFileSystemException {
+    byte[] bytes = new byte[len];
+    abfsClient.read(path.toUri().getPath(), 0, bytes, 0,
+        len, null, null, null,
+        tracingContext);
+    return bytes;
+  }
+
+  @VisibleForTesting
+  void createRenamePendingJson(Path path, byte[] bytes)
```

Comment:
why explicit calls to create, append and flush, why not use fs.create which returns OuputStream and just call write, hysnc and flush on that ?

Reply:
Thats what for I added the callbacks. We cannot have cycle of classes, right? I dont think client having a pointer to fileSystem would be good. What you feel?